### ISSUE TITLE
[BUG] 3.0.0 typescript compiler: Could not find a declaration file for module 'passport-strategy'

### ISSUE THREAD

**Issue by rskvazh** [node-saml/passport-saml#606]
**To Reproduce**

1. `npm install passport passport-saml @types/passport`
2. `tsc`

Got error: 

```shell
node_modules/passport-saml/lib/passport-saml/strategy.d.ts:1:46
error TS7016: Could not find a declaration file for module 'passport-strategy'. '/Users/rskvazh/Projects/SpatialChat/spatialchat-universe/node_modules/passport-strategy/lib/index.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/passport-strategy` if it exists or add a new declaration (.d.ts) file containing `declare module 'passport-strategy';`

1 import { Strategy as PassportStrategy } from "passport-strategy";
                                               ~~~~~~~~~~~~~~~~~~~
Found 1 error.
```

**Expected behavior**

`npm i --save-dev @types/passport-strategy` fixed the error and successfully compiled, but I think this should be fixed in a passport-saml package.

**Environment**

- Node.js version: v14.15.1
- passport-saml version: 3.0.0
- typescript: 4.3.2

**Comment by markstos** [comment]
The problem here was that you didn't include `--dev` in your original install statement. We already have the correct version of `@types/passport-strategy` as a dev dependency. 

The module is not needed for "production" installs, so it's correct that we don't have it as a production dependency.

**Comment by BigForNothing** [comment]
Can we reopen this? This is not the solution. @types/passport-strategy should be part of the dependencies, and not the devDependencies.

**Comment by cjbarth** [comment]
Types should always be installed as dev dependencies. See https://github.com/DefinitelyTyped/DefinitelyTyped#npm

**Comment by BigForNothing** [comment]
Their docs are (sort-of) wrong. See the official typescript website [documentation](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#dependencies) for publishing:

> Our package exposes declarations from each of those, so any user of our browserify-typescript-extension package needs to have these dependencies as well. For that reason, we used "dependencies" and not "devDependencies", because otherwise our consumers would have needed to manually install those packages. If we had just written a command line application and not expected our package to be used as a library, we might have used devDependencies.

The "sort-of" is because it should go in to `devDependencies` if you aren't publishing (which is what your link. If you are and someone is consuming it (like passport-saml), it goes into dependencies.

For passport-saml:
```
"dependencies": {
    "node-saml": "4.0.0-beta.1",
    "passport-strategy": "^1.0.0"
  },```
That should include @types/passport-strategy@^0.2.35 since it will be consumed. node-saml ships it's own declaration files though, so nothing extra to specify. 

If you look at node-saml's package.json:
```
"dependencies": {
    "@xmldom/xmldom": "^0.7.2",
    "debug": "^4.3.1",
    "xml-crypto": "^2.1.3",
    "xml-encryption": "^1.3.0",
    "xml2js": "^0.4.23",
    "xmlbuilder": "^15.1.1"
  },
``` 

These shold be included in those dependencies:

    "@types/debug": "^4.1.5",
    "@types/xml-crypto": "^1.4.1",
    "@types/xml-encryption": "^1.2.0",
    "@types/xml2js": "^0.4.8",

Basically the rule of thumb is "is the compiled code utilizing a @types package? If so, include it in dependencies"

**Comment by cjbarth** [comment]
I follow what you are saying, but I'm still not clear about your specific use-case. If you are using Passport and Passport Strategies with TypeScript, why wouldn't you include the types you need? Otherwise, what is deficient about the types that `passport-saml` exports?

It seems that a consumer of `passport-saml` shouldn't be using any version of, for example, `debug` obtained as a dependency of `passport-saml`, and thus shouldn't need any types provided by `passport-saml` for the `debug` module. If you project uses the `debug` module, then I would expect that you'd include both the `debug` and `@types/debug` packages. Otherwise `passport-saml` compiles down to JavaScript and so at runtime you wouldn't need any `debug` module types to use any of `passport-saml`s public APIs.

**Comment by BigForNothing** [comment]
> I follow what you are saying, but I'm still not clear about your specific use-case. If you are using Passport and Passport Strategies with TypeScript, why wouldn't you include the types you need? Otherwise, what is deficient about the types that `passport-saml` exports?

```
../../../node_modules/passport-saml/lib/passport-saml/strategy.d.ts:1:46 - error TS7016: Could not find a declaration file for module 'passport-strategy'. '/home/username/repos/backend/node_modules/passport-strategy/lib/index.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/passport-strategy` if it exists or add a new declaration (.d.ts) file containing `declare module 'passport-strategy';`

1 import { Strategy as PassportStrategy } from "passport-strategy";
```

I'm not the developer of our package that is using passport-saml. `@types/passport-strategy` triggered as "unused" by depcheck, which is why I even started looking into this. If `@types/passport-strategy` was included, this error wouldn't be a problem. It's been the [expectation for years](https://web.archive.org/web/20160928060100/https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html) that when publishing a package, it includes types needed to consume. I can definitely add another ignore item to depcheck, but I figured it was best to report it so it can be inline with community expectations.

> It seems that a consumer of `passport-saml` shouldn't be using any version of, for example, `debug` obtained as a dependency of `passport-saml`, and thus shouldn't need any types provided by `passport-saml` for the `debug` module. If you project uses the `debug` module, then I would expect that you'd include both the `debug` and `@types/debug` packages. Otherwise `passport-saml` compiles down to JavaScript and so at runtime you wouldn't need any `debug` module types to use any of `passport-saml`s public APIs.

`@types/debug` maybe be a bad example, because I assume you aren't shipping anything that is exposing that type. It's a rule of thumb to include the `@types` packages alongside the packages in dependencies. I don't have the time to check out the compiled code to see whats truly needed to be shipped.

**Comment by cjbarth** [comment]
Thank you for providing the error you are seeing. I'll investigate. It seems that we might be `extends`ing a `passport-strategy` type from `@types/passport-strategy`. In that case, we'll either need to do as you say, or fix our types.

**Comment by BigForNothing** [comment]
> especially not for those that don't use TypeScript
It benefits people who don't use TypeScript because their IDE will provide the relevant info still. 

Also, this is a TypeScript project and should follow the recommendations. I linked the info [previously](https://github.com/node-saml/passport-saml/issues/606#issuecomment-987264831) on the proper way to handle it. The proper thing is to fix the package.json to include the missing types for the dependencies. I can't think of any real reason not to do it. 

NPM also installs peer dependencies now too. So really, there isn't a good one in my mind to not include them. Maybe to upset yarn users who use TypeScript?
